src/autoDeleteMessages.ts

```diff
@@ -1,23 +1,55 @@
 import { BotContext } from './context';
 
-export interface AutoDeleteMessage {
-  [messageId: number]: number;
-}
+const expired = (createdAt: number, timeout: number) => {
+  const minute = 60000;
+  return Math.floor((Date.now() - createdAt) / minute) >= timeout;
+};
 
 export const runMessageRecycling = async (
   ctx: BotContext
 ): Promise<void> => {
-  const database = ctx.database;
+  const database = await ctx.loadDatabase();
   const messages = database.getAutoDeleteMessages();
-  const minute = 60000;
-  const recycleTimeout = 1;
-
-  Object.entries(messages).map(async ([id, createdAt]) => {
-    if (
-      Math.floor((Date.now() - createdAt) / minute) >= recycleTimeout
-    ) {
-      await ctx.deleteMessage(+id);
-      await database.removeAutoDeleteMessage(+id);
+  const messageTimeout = ctx.config.messageTimeoutInMinutes;
+
+  Object.entries(messages).map(async ([unsafeId, createdAt]) => {
+    if (!expired(createdAt, messageTimeout)) {
+      return;
+    }
+
+    const id = Number(unsafeId);
+    const chatId = ctx.chat?.id;
+
+    try {
+      ctx.logger.info(
+        `Trying to delete expired message with id "${id}" from the chat "${chatId}"`
+      );
+
+      await ctx.deleteMessage(id);
+
+      ctx.logger.info(
+        `Deleted expired message with id "${id}" from the chat "${chatId}"`
+      );
+    } catch (err) {
+      ctx.logger.warn(
+        `Failed to delete expired message with id "${id}" from the chat "${chatId}"`
+      );
+    }
+
+    try {
+      ctx.logger.info(
+        `Trying to delete expired message with id "${id}" from the database`
+      );
+
+      await database.removeAutoDeleteMessage(id);
+
+      ctx.logger.info(
+        `Deleted expired message with id "${id}" from the database`
+      );
+    } catch (err) {
+      ctx.logger.warn(
+        `Failed to delete expired message with id "${id}" from the database`
+      );
     }
```

Comment:
I think an extra step is needed here for some resilience.. Probably checking if the message still exists in the chat before trying to delete it makes sense.

In case it doesn't, we can simply remove it from our database, because it can happen that someone deletes their own message or another admin for example.

So the process would be something like:

- check if message exists in the chat
- if exists: try to delete it from the chat and then delete it from the database
- if it doesn't: just delete it from the database

Reply:
just checked telegram API and looks like there is no way to fetch a message from the chat through the id :(

Reply:
I added a retrial logic for the delete chat command, which makes it slightly more resilient in case of failures. The flow feels a bit more stable now, although the code is not as pretty as before hehe